### PR TITLE
Phase 6: Reynolds Number Perturbation Augmentation — OOD-Re Robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The gap/stagger perturbation augmentation (PR #2115, σ=0.02) was our biggest recent win: **p_oodc -4.9%, p_re -1.5%**. The mechanism is domain randomization — training with slightly perturbed geometry features teaches the model to be robust to OOD geometry.

We apply the exact same paradigm to Reynolds number. The `val_ood_re` split tests at **Re=4.445M (fully OOD)**. By adding small Gaussian noise to `log_Re` (feature index 13 in the 24-dim input) during training, we teach the model that flows at slightly different Re values should yield smoothly-varying predictions — directly targeting p_re robustness.

**Why this should work:**
- CFD flows vary smoothly with Re (Navier-Stokes equations are smooth in Re)
- Adding noise σ on log_Re = training on a distribution of Re values within ~e^σ of the true Re, giving implicit coverage of nearby Re values
- Gap/stagger aug proves domain randomization works for geometry parameters; Re is the natural next axis to regularize
- The log_Re feature is global (same value for all N nodes per sample), making implementation trivial: perturb `x[:, :, 13]` once per batch

**Primary target: p_re < 6.45. Secondary: p_oodc < 7.92.**

## Instructions

In `cfd_tandemfoil/train.py`, add a new config flag and augmentation block following the exact same pattern as `aug_gap_stagger_sigma`:

**1. Add config field** (near line ~961, after `aug_gap_stagger_sigma`):
```python
aug_re_sigma: float = 0.0  # std of Gaussian noise on log_Re feature (index 13) during training
```

**2. Add augmentation block** in the training loop, immediately after the gap/stagger aug block (~line 1555), and BEFORE the standardization call `x = (x - stats["x_mean"]) / stats["x_std"]`:
```python
# Reynolds number perturbation augmentation (all samples)
if cfg.aug_re_sigma > 0.0:
    _B = x.size(0)
    _re_noise = torch.randn(_B, device=x.device) * cfg.aug_re_sigma  # [B]
    x[:, :, 13] = x[:, :, 13] + _re_noise.unsqueeze(1)  # broadcast to all N nodes
```

**Implementation notes:**
- Apply BEFORE `x = (x - stats["x_mean"]) / stats["x_std"]` — noise is in raw feature space, same as gap/stagger aug
- Apply to ALL samples (not tandem-only) — Re is a universal feature
- No clamping needed — small Gaussian perturbation on log_Re is always physically valid
- Training-only (no change to eval/validation code)
- Optional sanity metric: log `aug/re_noise_std` as the actual std of Re noise per epoch

**Experiment sweep — run 2 seeds for each sigma value:**

```bash
# σ=0.05 (start here — analogous scale to gap/stagger σ=0.02 on log_Re range)
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/re-aug-s05-s42" --wandb_group phase6/re-perturb-aug \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 \
  --aug_re_sigma 0.05
```

Run seeds 42 and 43 for σ=0.05 first. If both look promising (p_re improves), also run σ=0.10 and σ=0.20 with the same wandb_group. If σ=0.05 degrades p_tan by >2%, stop and report.

## Baseline

Current single-model baseline (PR #2104 + PR #2115 merged, 8-seed mean, seeds 42-49):

| Metric | 8-seed mean | Target to beat |
|--------|-------------|----------------|
| p_in   | 13.19 ± 0.33 | < 13.19 |
| p_oodc | 7.92 ± 0.17  | < 7.92  |
| p_tan  | 30.05 ± 0.36 | < 30.05 |
| p_re   | **6.45 ± 0.07** | **< 6.45** |

Per-seed baselines for direct 2-seed comparison:
- Seed 42: p_in≈13.4, p_oodc≈7.7, p_tan≈29.8, p_re≈6.4
- Seed 43: p_in≈13.2, p_oodc≈7.8, p_tan≈29.3, p_re≈6.4

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Reynolds Number Perturbation Augmentation (σ=0.05)

### Configuration
- `--aug_re_sigma 0.05` with all baseline flags (`--aft_foil_srf --aug_gap_stagger_sigma 0.02`)
- Seeds 42 and 43, 155 epochs reached (180-min timeout)
- W&B group: `phase6/re-perturb-aug`

### Results

| Metric | Baseline s42 | σ=0.05 s42 | Δ | Baseline s43 | σ=0.05 s43 | Δ |
|--------|-------------|------------|---|-------------|------------|---|
| p_in   | ~13.4       | 13.4       | ~0 | ~13.2       | 13.1       | ~-0.1 |
| p_tan  | ~29.8       | 30.2       | +0.4 | ~29.3       | 30.8       | **+1.5** |
| p_oodc | ~7.7        | 7.8        | +0.1 | ~7.8        | 7.7        | -0.1 |
| p_re   | ~6.4        | 6.5        | +0.1 | ~6.4        | 6.5        | +0.1 |

### W&B Runs
- Seed 42: `ecxe1ti3` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ecxe1ti3
- Seed 43: `29xvq9bz` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/29xvq9bz

### Analysis

**Negative result.** σ=0.05 Re perturbation did not improve p_re (6.5 vs baseline 6.4 for both seeds) and caused p_tan regression (+1.5 on seed 43). Per the PR instructions, since σ=0.05 did not improve p_re, I did not proceed with σ=0.10 or σ=0.20.

**Why this likely didn't work:**
1. **log_Re is already normalized** — after standardization, the feature range is compressed, so σ=0.05 in raw space may translate to a very small perturbation relative to the feature's std. The noise may be too small to provide meaningful regularization.
2. **OOD-Re gap may be too large** — the training Re values are log(Re)≈14.5-15.4 while OOD Re=4.445M maps to log(Re)≈15.3. The OOD-Re split may not be truly "far" OOD in log-space, so there's limited headroom for augmentation to help.
3. **p_tan regression** suggests the Re noise interferes with the model's ability to learn tandem-specific pressure patterns, possibly because Re and geometry features interact non-linearly in tandem configurations.

### Suggested Follow-ups
- **Targeted Re aug (OOD-Re only):** Instead of perturbing all samples, only perturb samples near the OOD-Re boundary to extend coverage in that direction
- **Conditional Re aug:** Apply different sigma values for single vs tandem samples, since the tandem regression suggests they need different treatment
- **Feature-space noise** (post-normalization): Apply noise in standardized space where σ=0.05 would be more meaningful relative to the feature distribution

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2125 because ADVISOR: Closing as dead end. Results clearly show no improvement over baseline:

**2-seed avg:** p_in≈13.25 (baseline 13.24), p_tan≈30.5 (baseline 30.53), p_oodc≈7.75 (baseline 7.73), p_re≈6.50 (baseline 6.50). No metric beats baseline.

Seed 43 showed a significant p_tan regression (+1.5 to 30.8) indicating Re noise actively interferes with tandem pressure learning. The student's analysis is spot-on: log_Re has a narrow range in training data, so small perturbations in raw space provide negligible regularization, while larger perturbations would move far from physical reality.

**Conclusion:** Reynolds number perturbation augmentation does not help OOD-Re robustness in this problem. The OOD-Re split (Re=4.445M) is apparently not far enough in log-space to benefit from domain randomization. Adding Re-perturbation to the confirmed dead-ends list.

Good work on the analysis, thorfinn. Your suggested follow-ups (targeted Re aug, conditional Re aug, feature-space noise) would all face the same fundamental issue — the OOD-Re gap is simply too small in log-space. Moving you to a higher-impact direction next..
